collector/controller.go

```diff
@@ -0,0 +1,64 @@
+package collector
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	k8sscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	pipelinev1beta1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+)
+
+func NewManager(cfg *rest.Config, options ctrl.Options, logger log.Logger) (ctrl.Manager, error) {
+	// we have seen in testing that this path can get invoked prior to the PipelineRun CRD getting generated,
+	// and controller-runtime does not retry on missing CRDs.
+	// so we are going to wait on the CRDs existing before moving forward.
+	apiextensionsClient := apiextensionsclient.NewForConfigOrDie(cfg)
+	if err := wait.PollImmediate(time.Second*5, time.Minute*5, func() (done bool, err error) {
+		_, err = apiextensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), "pipelineruns.tekton.dev", metav1.GetOptions{})
+		if err != nil {
+			level.Info(logger).Log("msg", fmt.Sprintf("get of taskrun CRD failed with: %s", err.Error()))
+			return false, nil
+		}
+		level.Info(logger).Log("msg", "get of pipelinerun CRD returned successfully")
+		return true, nil
+	}); err != nil {
+		level.Error(logger).Log("msg", "timed out waiting for taskrun CRD to be created", err)
+		return nil, err
+	}
+
+	options.Scheme = runtime.NewScheme()
+	if err := k8sscheme.AddToScheme(options.Scheme); err != nil {
+		return nil, err
+	}
+	if err := pipelinev1beta1.AddToScheme(options.Scheme); err != nil {
```

Comment:
Please add a TODO here - we should upgrade to Tekton v1 now that the API is available.

Reply:
Sure I can do that, though I would think converting from v1beta1 to v1 is going to be a product wide Jira based thing.